test/impute.jl

```diff
@@ -0,0 +1,124 @@
+            @test isequal(r[k], v)
+        end
+    end
+
+    @testset "impute" begin
+        expected = KeyedDataset(
+            flatten([
+                :train => [
+                    :temp => KeyedArray(allowmissing([1.0 1.1; 2.2 2.2; 3.0 3.3]); time=1:3, id=[:a, :b]),
+                    :load => KeyedArray(allowmissing([7.0 7.7; 8.0 8.0; 9.0 9.9]); time=1:3, loc=[:x, :y]),
+                ],
+                :predict => [
+                    :temp => KeyedArray(allowmissing([1.0 1.0; 2.0 2.2; 3.0 3.3]); time=1:3, id=[:a, :b]),
+                    :load => KeyedArray(allowmissing([7.0 7.7; 8.1 8.1; 9.0 9.9]); time=1:3, loc=[:x, :y]),
+                ]
+            ])...
+        );
+
+        r = Impute.substitute(ds; dims=:time)
+        for (k, v) in expected.data
+            @test isequal(r[k], v)
+        end
+
+        expected = KeyedDataset(
+            flatten([
+                :train => [
+                    :temp => KeyedArray([1.0 1.1; missing 2.2; 3.0 3.3]; time=1:3, id=[:a, :b]),
+                    :load => KeyedArray(allowmissing([7.0 7.7; 8.0 8.8; 9.0 9.9]); time=1:3, loc=[:x, :y]),
+                ],
+                :predict => [
+                    :temp => KeyedArray([1.0 missing; 2.0 2.2; 3.0 3.3]; time=1:3, id=[:a, :b]),
+                    :load => KeyedArray(allowmissing([7.0 7.7; 8.1 8.8; 9.0 9.9]); time=1:3, loc=[:x, :y]),
+                ]
+            ])...
+        );
+
+        r = Impute.substitute(ds; dims=:loc)
+        for (k, v) in expected.data
+            @test isequal(r[k], v)
+        end
+    end
+
+    @testset "filter" begin
+        # All time axis are constraint, so they must match and removing a row from 1
+        # component will remove it from all
+        expected = KeyedDataset(
+            flatten([
+                :train => [
+                    :temp => KeyedArray([3.0 3.3]; time=3:3, id=[:a, :b]),
+                    :load => KeyedArray([9.0 9.9]; time=3:3, loc=[:x, :y]),
+                ],
+                :predict => [
+                    :temp => KeyedArray([3.0 3.3]; time=3:3, id=[:a, :b]),
+                    :load => KeyedArray([9.0 9.9]; time=3:3, loc=[:x, :y]),
+                ]
+            ])...
+        );
+
+        r = Impute.filter(ds; dims=:time)
+        for (k, v) in expected.data
+            @test isequal(r[k], v)
+        end
+
+        # Only :load has a shared :loc axis, so we see that that :y location is dropped from
+        # both train and predict.
+        expected = KeyedDataset(
+            flatten([
+                :train => [
+                    :temp => KeyedArray([1.0 1.1; missing 2.2; 3.0 3.3]; time=1:3, id=[:a, :b]),
+                    :load => KeyedArray([7.0 8.0 9.0]'; time=1:3, loc=[:x]),
+                ],
+                :predict => [
+                    :temp => KeyedArray([1.0 missing; 2.0 2.2; 3.0 3.3]; time=1:3, id=[:a, :b]),
+                    :load => KeyedArray([7.0 8.1 9.0]'; time=1:3, loc=[:x]),
```

Comment:
`[7.0 8.1 9.0]'` seems like a really ugly way to create a 3x1 matrix... anyone know of a cleaner way to write that?